Alex and Happy Part 1/article.md

```diff
@@ -0,0 +1,817 @@
+# Parsing with Haskell (Part 1): Lexing with Alex
+
+<!--
+* [X] generic introduction
+-->
+
+This is the first part of the two parts _Parsing with Haskell_ series.
+Looking for the second part instead? You can find it [here](https://serokell.io/blog/parsing-with-happy).
+
+A critical part of the majority of the compilers is grammar.
+The grammar exposes the compiler's front-end as a convenient way to build programs.
+
+This two-parts tutorial will look into two tools often used together by Haskellers to parse programs, namely Alex and Happy.
+
+Both Alex and Happy are industrial-strength tools powerful enough to parse even Haskell itself.
+At the bottom of this tutorial, you will find links to GHC's lexer and parser if you are curious to see how they look.
+
+This tutorial was written using GHC version 9.0.2, Stack resolver LTS-19.8, Alex version 3.2.7.1, and Happy version 1.20.0.
+
+## Our grammar: MiniML
+
+<!--
+* [X] describe what we support
+* [X] example snippet
+-->
+
+For this tutorial, we will introduce a small grammar based on [ML](https://en.wikipedia.org/wiki/ML_(programming_language)), which we will call MiniML (read as "minimal").
+As the name suggests, it should introduce a minimal syntax to get you started with Alex and Happy. Still, at the same time, we hope to introduce enough concepts so that you will be able to create a useful grammar for your programming language and use these tools effectively.
+
+Some of the features that this tutorial series will support are:
+* Variables, constants, function declarations, function applications, local bindings, and operators.
+* Some simple types, as well as arrow types (functions).
+* Conditional expressions: if-then-else and if-then.
+* Block comments.
+* The reader will be invited in the proposed exercises to extend the language with line comments and rational numbers. Later, in the second part, exercises will introduce patterns, list access, and pattern matching.
+
+The snippet below demonstrates a program written in MiniML:
+
+```sml
+let the_answer : int =
+  let a = 20 in
+  let b = 1 in
+  let c = 2 in
+  a * c + b * c
+
+let main (unit : ()) : () =
+  print ("The answer is: " + the_answer)
+```
+
+## Lexing
+
+Before we can start parsing, we should first write a lexer to the grammar, which is also known as _lexical analyzer_, _scanner_, and _tokenizer_.
```

Comment:
Is it known as all of those things or either of those things? Might need or instead of and at the end of sentence.